File: 0249. Group Shifted Strings.java
// Medium
// HashTable, String
// Time:O(N*K),Space:O(N*K)
// https://leetcode.cn/problems/group-shifted-strings/

import java.util.*;

class Solution {
    public List<List<String>> groupStrings(String[] strings) {
        Map<String, List<String>> map = new HashMap<>();
        for (String s : strings) {
            // calculate the diff of each string
            StringBuilder sb = new StringBuilder();
            for (int i = 1; i < s.length(); i++) {
                int diff = (s.charAt(i) - s.charAt(i - 1) + 26) % 26;
                sb.append(diff).append(",");
            }

            // use diff as key, string as value
            String key = sb.toString();
            if (!map.containsKey(key)) {
                map.put(key, new ArrayList<>());
            }
            map.get(key).add(s);
        }
        return new ArrayList<>(map.values());
    }
}

// 判断每个字符串是否是shifted版本的另一个字符串
// 例如 "abc"和"zab"的字符差分别为：
// a->z的差为-1或25， b->a的差为-1或25，c->b的差为-1或25
// 可以计算字符串中相邻字符的差值，并将其标准化mod26，判断是否为相同的移位版本
// 再使用哈希表将标准化后的差值作为key，相同差值的字符串为value归类为一组

File: 0001-0500/0249. Group Shifted Strings.py
# Time:O(n*k), Space:O(n*k)

from typing import List

class Solution:
    def groupStrings(self, strings: List[str]) -> List[List[str]]:
        groups = {}

        for s in strings:
            key = []
            for i in range(len(s) - 1):
                diff = (ord(s[i + 1]) - ord(s[i])) % 26
                key.append(diff)
            
            key = tuple(key)
            if key not in groups:
                groups[key] = []
            groups[key].append(s)
        
        return list(groups.values())
    